streams/shared/src/main/scala/scalaz/stream/ZSink.scala

```diff
@@ -602,12 +708,110 @@ object ZSink {
         }
     }
 
+  /**
+   * Collects incoming elements into a list as long as they verify predicate `p`.
+   *
+   * TODO rename to `collectWhile`/`accumWhile`?
+   */
+  def readWhile[A](p: A => Boolean): ZSink[Any, Nothing, A, A, List[A]] =
+    readWhileM(a => IO.succeed(p(a)))
+
+  /**
+   * Collects incoming elements into a list as long as they verify effectful predicate `p`.
+   *
+   * TODO rename to `collectWhileM`/`accumWhileM`?
+   */
+  def readWhileM[R, E, A](p: A => ZIO[R, E, Boolean]): ZSink[R, E, A, A, List[A]] =
+    ZSink
+      .foldM[R, E, A, A, List[A]](ZIO.succeed(List.empty[A])) { (s, a: A) =>
+        p(a).map(if (_) Step.more(a :: s) else Step.done(s, Chunk(a)))
+      }
+      .map(_.reverse)
+
+  /**
+   * TODO I don't understand why these combinators are invariant in all types
```

Comment:
Ah the name is misleading. It's just that these combinators only work when `A0 =:= A`.

Reply:
I wonder how this is going to look in the generated scaladoc. I’m wondering whether it would be clearer if these methods were moved into `ZSink` with an implicit proof?

Reply:
It would! And in fact, that's the standard for ZIO's methods.

Reply:
Great, I will do the moving then.

Reply:
Done ✌️